### ISSUE TITLE
Discussion: Impersonation/Masquerade feature

### ISSUE THREAD

**Issue by LordSimal** [cakephp/authentication#473]
As a member of the CakePHP discourse already askeded here https://discourse.cakephp.org/t/cakephp-4-x-impersonate-login-as-different-user/9559 an impersonation/masquerade feature would be very useful.

raul338 already presented a solution with a custom Authenticator based on the SessionAuthenticator here: https://discourse.cakephp.org/t/cakephp-4-x-impersonate-login-as-different-user/9559/3

The question now is if such a impersonation logic will always have to be implemented per user since the logic **which user can impersonate** can differ widely or if this can be implemented in this plugin (or an extra plugin)?

Was just curious what your opinions on that would be.

**Comment by markstory** [comment]
I like the general direction of those changes. If we added an impersonation feature to the plugin I would like to see it integrated more robustly though.

How would other persistent backends implement this? Currently the example directly writes to the session. Instead I would prefer we had an interface like:

```php
// In a controller
$this->Authorization->impersonate($user)
```

By having this more abstract interface we could enable other persistent backends provide the same behavior. The other persistent adapters we provide is the `CookieAuthenticator`. A similar key could work there, but I've not thought through all the implications it could create.

**Comment by steinkel** [comment]
While working on the implementation for the impersonation feature we were discussing about CookieAuthenticator. In all the cases we typically use this feature, we never assume the impersonation should survive the session. Even if we could implement CookieAuthenticator to also store the impersonation related data, or add a new cookie with the impersonation data, we think this case will be very rare. So we would NOT be implementing the CookieAuthenticator to store the impersonated user or data, if you loose the session, you will need to impersonate another user again.

What are your thoughts on this?

Thanks,

**Comment by LordSimal** [comment]
I would also agree that session based impersonation is the main focus here.

**Comment by jeremyharris** [comment]
Fwiw, my personal implementation of impersonation is session only and so far it's never been expected to survive past a session. 

It does feel buggy, however, when the session expires during impersonation:
- user A impersonates B
- visit page only B has permission to
- session expires
- the page reloads, authenticates with user A cookie
- error: user A cannot view this page 

One could argue this is a messaging problem or an application problem, I'm just bringing it up to share my experience.

**Comment by markstory** [comment]
I'm fine with session only impersonation.

**Comment by markstory** [comment]
Implemented in #532
